client/asset/dcr/dcr.go

```diff
@@ -4527,3 +4546,178 @@ func float64PtrStr(v *float64) string {
+		}
+		if !abandonTx {
+			if tx.Confirmations >= reqiredRedeemConfirms {
+				deleteMempoolRedeem = true
+			}
+			return &asset.ConfirmRedemptionStatus{
+				Confs:  uint64(tx.Confirmations),
+				Req:    reqiredRedeemConfirms,
+				CoinID: coinID,
+			}, nil
+		}
+	}
+
+	if !abandonTx && !errors.Is(err, asset.CoinNotFoundError) {
+		return nil, fmt.Errorf("problem searching for redemption transaction %s: %w", txHash, err)
+	}
+
+	// Redemption transaction is missing from the point of view of our node!
+	// Unlikely, but possible it was redeemed by another transaction. We
+	// assume a contract past its locktime cannot make it here, so it must
+	// not be refunded. Check if the contract is still an unspent output.
+
+	txHash, vout, err := decodeCoinID(redemption.Spends.Coin.ID())
+	if err != nil {
+		return nil, err
+	}
+
+	var spent bool
+	_, err = dcr.wallet.UnspentOutput(dcr.ctx, txHash, vout, 0)
+	if err != nil {
+		if errors.Is(err, asset.CoinNotFoundError) {
+			spent = true
+		} else {
+			return nil, err
+		}
+	}
+
+	if spent {
+		// First find the block containing the output itself.
+		scriptAddr, err := stdaddr.NewAddressScriptHashV0(redemption.Spends.Contract, dcr.chainParams)
+		if err != nil {
+			return nil, fmt.Errorf("error encoding contract address: %w", err)
+		}
+		_, pkScript := scriptAddr.PaymentScript()
+		outFound, block, err := dcr.externalTxOutput(dcr.ctx, newOutPoint(txHash, vout),
+			pkScript, time.Now().Add(-60*24*time.Hour)) // search up to 60 days ago
+		if err != nil {
+			return nil, err // possibly the contract is still in mempool
+		}
+		// Try to find a transaction that spends it.
+		spent, err := dcr.isOutputSpent(dcr.ctx, outFound)
+		if err != nil {
+			return nil, fmt.Errorf("error checking if contract %v:%d is spent: %w", *txHash, vout, err)
+		}
+		if spent {
+			vin := -1
+			spendTx := outFound.spenderTx
+			for i := range spendTx.TxIn {
+				sigScript := spendTx.TxIn[i].SignatureScript
+				sigScriptLen := len(sigScript)
+				if sigScriptLen < dexdcr.SwapContractSize {
+					continue
+				}
+				// The spent contract is at the end of the signature
+				// script. Lop off the front half.
+				script := sigScript[sigScriptLen-dexdcr.SwapContractSize:]
+				_, _, _, sh, err := dexdcr.ExtractSwapDetails(script, dcr.chainParams)
+				if err != nil {
+					// This is not our script, but not
+					// necessarily a problem. Ignoring the error.
+					continue
+				}
+				if bytes.Equal(sh[:], secretHash[:]) {
+					vin = i
+					break
+				}
+			}
+			if vin >= 0 {
+				_, height, err := dcr.wallet.GetBestBlock(dcr.ctx)
+				if err != nil {
+					return nil, err
+				}
+				hash := spendTx.TxHash()
+				return &asset.ConfirmRedemptionStatus{
+					Confs:  uint64(height - block.height),
+					Req:    reqiredRedeemConfirms,
+					CoinID: toCoinID(&hash, uint32(vin)),
+				}, nil
+			}
+		}
+		dcr.log.Warnf("Contract coin %v spent by someone but not sure who.", redemption.Spends.Coin.ID())
```

Comment:
Any point in examining the output address(es) for the tx to see if it's us?

Reply:
I know you're asking @JoeGruffins, but I have this question some thought in context of the recently-added refund detection code (https://github.com/decred/dcrdex/pull/2011). Ultimately, I decided the client primarily needs a path forward instead of retrying and failing forever.  If we did detect it was paid to address X, we could log it, maybe even try to see if wallet owns the address, but aside from detecting if we were hacked, I didn't see it being helpful.  User can bring up the tx on an explorer, so maybe we log the located refund txid?

Reply:
Good points. A good log message is probably the best we can do. This will probably never happen anyway. No change needed.

Reply:
Leaving as is for now.